### ISSUE TITLE
Research: compile frontier-authored operation procedures into reusable guarded macros

### ISSUE THREAD

**Issue by Unjuno** [Unjuno/agent-interface#857]
## Motivation

Agent Interface already has several adjacent mechanisms:

- #56 — compile grounded GUI understanding into reusable symbolic methods;
- #696 — represent bounded local execution as declarative sensorimotor graphs;
- #732 — install temporary resident reactive programs during frontier inference;
- #751 — retrieve and reuse retained interface skills;
- #9 / #853 — local learned System-1 policies for repeated typed decisions.

A missing architectural question sits before the learned-policy path:

> When Astra / a frontier model successfully reasons through an operation procedure, can Agent Interface compile the repeated stable part of that procedure into the **smallest reusable guarded macro**, instead of either replaying the frontier model on every use or prematurely training a learned local policy?

The intended hierarchy is not “everything becomes a local model.” The default should be:

```text
frontier reasoning
    -> fixed procedure if possible
    -> parameterized/conditional macro if sufficient
    -> deterministic rule/graph if needed
    -> tiny learned policy only for residual repeated decisions
    -> UNKNOWN / YIELD when the compiled artifact is not applicable
```

This issue records that compiler/lifecycle question only. It does **not** authorize a new experiment, runtime mutation, model call, GUI action, or repository implementation by itself.

## Distinction from adjacent issues

This issue should remain separate only if the research object is specifically **procedure-to-macro compilation and promotion**.

- #56 owns persistent symbolic GUI interfaces and end-to-end cold/warm economics.
- #696 owns declarative graph representation and static graph validity.
- #732 owns concurrent resident execution during model inference.
- #751 owns search/retrieval/consolidation after a skill already exists.
- #9/#853 own learned local-policy acceleration.

If implementation reduces to merely serializing a #696 graph, fold it into #696. If it is only retaining a #56 symbolic method, fold it into #56. The distinct claim here is that frontier-authored procedures can be **abstracted, parameterized, guarded, and promoted into a simpler reusable macro artifact before learned policy is considered**.

## Candidate compilation flow

```text
Astra / frontier model
    understands current task
    authors or executes a successful bounded procedure
             ↓
Procedure compiler
    identify stable steps
    identify parameters
    identify observable branch predicates
    identify preconditions / invalidation dependencies
    identify effect postconditions
    identify UNKNOWN / YIELD exits
             ↓
Reusable macro artifact
    typed inputs
    bounded deterministic branches
    current-evidence guards
    authority requirements
    explicit release / cleanup
    effect verification
    no authority of its own
             ↓
Ordinary deterministic admission/runtime
```

A macro is therefore not just a recorded click sequence. Candidate shape:

```text
CompiledMacro {
    macro_id
    version
    semantic_intent

    typed_parameters
    observable_preconditions
    steps / bounded branches
    current-evidence dependencies
    authority requirements
    expected_effects
    forbidden_effects

    max_duration
    max_transitions
    max_actuator_events

    UNKNOWN / YIELD exits
    invalidation conditions
    release_plan

    provenance
    positive evidence refs
    retained negative evidence refs
}
```

Historical observation, a prior successful coordinate, or frontier confidence is never execution authority.

## Macro-first backend selection principle

Before allocating a learned backend for a repeated operation, ask whether the repeated behavior can be represented adequately by a simpler artifact.

Candidate ordering:

```text
OPEN_LOOP_MACRO
    fixed deterministic action sequence

PARAMETRIC_MACRO
    same sequence with typed arguments

CONDITIONAL_MACRO
    bounded observable branches / assertions / retries

DETERMINISTIC_RULE_OR_GRAPH
    richer state-dependent control

TINY_LEARNED_POLICY
    repeated decision boundary remains expensive to hand-author

HEAVIER_LOCAL_MODEL
    only after smaller forms fail their declared task envelope

YIELD
    semantic novelty / ambiguity / unsupported state
```

This ordering is a research heuristic, not an unconditional production rule. A simple macro is preferable only when it preserves correctness and does not hide relevant state variation.

## Procedure abstraction, not raw replay

A frontier execution trace such as:

```text
click absolute point
enter literal value
click another absolute point
```

should not automatically become the retained artifact.

The compiler should attempt to recover the reusable relation:

```text
activate(email_field)
edit(value=email)
activate(submit)
verify(submitted)
```

with explicit target/current-evidence requirements.

The core scientific question is whether the stable procedure can be recovered without smuggling task-specific hidden state or converting historical coordinates into current authority.

## Promotion conditions

A frontier-authored procedure is a candidate for macro promotion only when all of the following can be made explicit:

1. **Typed input boundary** — variable inputs/targets are named rather than baked into the trace.
2. **Observable applicability** — the runtime can decide whether the macro is applicable from allowed current evidence.
3. **Bounded control flow** — branches/retries are finite and auditable.
4. **Mechanical success condition** — the macro has an observable terminal/effect contract.
5. **Fail-closed uncertainty** — unknown, ambiguity, missing target/evidence or violated precondition yields rather than guessing.
6. **Authority separation** — the macro cannot grant, extend, revive or reinterpret authority.
7. **Release/cleanup** — every owned key/button/contact/resource has deterministic terminal cleanup semantics.
8. **Negative evidence preservation** — a known failing envelope is retained and cannot be erased by later successful variants or consolidation.

One successful frontier episode alone is not enough evidence for broad retention.

## H — falsifiable architectural hypothesis

For a class of repeated frontier-authored operation procedures, the stable reusable portion can be compiled into a typed guarded macro that reproduces the intended independently verified effect on fresh in-envelope cases with fewer frontier boundaries than repeated frontier execution, while preserving UNKNOWN/YIELD outside its proven envelope.

A second hypothesis is that a **macro-first** compilation ladder reduces unnecessary learned-policy allocation: some repeated tasks that appear to need a local policy will be adequately handled by parameterized or conditional macros.

Neither hypothesis is tested by creating this issue.

## T — future minimum discriminator, not currently authorized

When this issue is eventually activated, the smallest useful test should use one already-understood bounded desktop workflow and compare only the execution representation while keeping the semantic task, authority, current-evidence checks and effect scorer fixed.

Candidate future arms:

1. `FRONTIER_REPEAT` — frontier authors/executes the procedure each time;
2. `COMPILED_MACRO` — a macro compiled from an earlier frontier-authored successful procedure;
3. optional `TINY_POLICY` only if the macro fails for a specifically identified repeated state-dependent reason.

The first test should not optimize latency or model quality broadly. It should ask whether procedure abstraction is correct and whether UNKNOWN/YIELD is preserved under a fresh applicability break.

Before any such execution: source-first freeze exact procedure/compiler/macro schema/cases/scorer; check current Issue/PR ownership; retain first outcome; change one material factor only.

## D — future decision rule

A future scoped PASS should require all of:

- fresh in-envelope cases preserve the independent task/effect contract;
- zero wrong-target / stale-authority admission attributable to compiled state;
- the macro never executes on UNKNOWN, missing required current evidence, or out-of-envelope state;
- every execution remains inside ordinary deterministic authority and release semantics;
- the compiled artifact is materially simpler than repeated frontier execution and does not merely embed another general planner;
- at least one repeated frontier boundary is actually eliminated or avoided;
- known negative predecessor evidence remains unchanged and visible.

`HOLD` if the procedure cannot be abstracted without making applicability or effect semantics ambiguous.

`REJECT_MACRO_COMPILATION_FOR_TASK_CLASS` if a bounded macro repeatedly fails because materially different states require decisions that cannot be represented by the frozen finite branch/guard structure without semantic weakening. Only then should a learned-policy successor be considered.

## C — competing explanations / failure modes

- Apparent reuse may come from a fixture-specific target locator rather than procedure compilation.
- A “compiled macro” may simply hide a large hand-authored program and shift complexity out of view.
- Branch explosion may make the macro harder to validate than asking the frontier model again.
- Historical successful coordinates may accidentally become stale execution state.
- Applicability checking may cost as much as fresh frontier reasoning for low-reuse tasks.
- A learned tiny policy may be simpler than enumerating a genuinely irregular decision surface.
- Reusing a macro can preserve a locally correct but globally stale strategy; semantic novelty must still YIELD.

## U — scope and uncertainty

This issue does not establish:

- that all frontier behavior should be compiled;
- that macros dominate learned local policies;
- that compiled procedures are portable across applications;
- that one successful demonstration is sufficient for skill retention;
- any latency/token/cost gain before measured end-to-end accounting;
- any Jev dependency or integration.

Jev may remain an external architectural comparator for the broad idea of compiling expensive reasoning into cheaper local execution, but this issue must not depend on or integrate Jev.

## Relationship to skill retention

If a compiled macro later earns retained evidence, #751 may index it as a reusable skill. Registry retrieval must remain separate from execution authority:

```text
retrieved macro
    != executable authority
```

Current applicability, target/evidence dependencies, ordinary authority admission and effect verification still occur at use time.

## Measurement direction when eventually activated

Do not collapse different work dimensions into one score. Record separately where applicable:

- frontier boundaries avoided per successful use;
- compilation/validation cost;
- warm local execution cadence/cost;
- local observations and deterministic checks;
- correctness / wrong-target / stale-action outcomes;
- UNKNOWN/YIELD frequency;
- authority/release correctness;
- repair/invalidation cost;
- reuse count and break-even;
- transfer to a second task/domain only after the first scoped mechanism result.

## Current stop

Issue creation only. No experiment, branch, source freeze, model call, GUI action, runtime mutation, PR, or promotion is authorized by this issue now.

Next discriminator when this lane is intentionally activated:

> Can one real frontier-authored bounded procedure be reduced to a typed conditional macro with explicit current-evidence guards and UNKNOWN/YIELD, without turning the macro into a hidden planner?

**Comment by Unjuno** [comment]
## Research intake — training-free local specialization and latency boundaries (2026-09-17)

**Status: source/literature review and design clarification only.** No experiment, model download/load/training, benchmark allocation, GUI action, branch, PR, or runtime change is started by this note. The existing issue-only stop remains in force. Jev remains an external comparator, not a dependency or implementation source.

### Question refined by the user

Do not treat a hosted decision API's response time as the lower bound for a custom local decision layer. Astra may specialize a repeated decision into a small executable artifact; where semantic interpretation remains necessary, an already-pretrained small local classifier might be usable without task-specific weight training.

Separate two meanings of “no training”:

- **Model-free specialization:** Astra authors bounded rules, parameterized macros, finite branches or fixed geometric/scoring logic. No learned weights are required. Authoring, validation and maintenance still have costs.
- **No task-specific training:** use frozen, previously trained weights for zero-shot candidate scoring. This is not an untrained model and does not inherit correctness/calibration merely from being small.

### Primary-source findings, not local measurements

1. [TypeSafe's 2026-09-15 Jev introduction](https://typesafe.ai/blog/introducing-system-one-models-and-jev) reports **70–500 ms end-to-end**, with published evaluations generally run from West Coast laptops near its service. The consulted material does not identify the model parameter count or a network/queue/server-compute breakdown. Thus “200 ms is mainly network” and “Jev itself would take under 60 ms locally” remain unverified. Local CPU deployment could also replace faster server hardware.
2. [GLiClass-Edge-v3.0 model card](https://huggingface.co/knowledgator/gliclass-edge-v3.0) provides a concrete independent optional candidate: **32.7M parameters, 131 MB listed weight size, Apache-2.0**, zero-shot text/label classification in one forward pass. It is pretrained; the card also describes prior logical-task fine-tuning. This is an optional text-decision leaf, not a screenshot-grounding model or demonstrated Jev equivalent.
3. [GLiClass paper, arXiv:2508.07662](https://arxiv.org/html/2508.07662v1) evaluates speed on **one NVIDIA A6000, batch 1**, 64/256/512-token texts and 1–128 labels, ten forward passes per configuration. Reported edge throughput across label counts is **82.64–103.81 examples/s**, averaged across text lengths. This is not a CPU or p95 result. Its classification results also expose a quality cost from selecting the smallest model. No CPU-60-ms or computer-use safety claim follows.
4. [Official GLiClass.c](https://github.com/Knowledgator/GLiClass.c) documents an ONNX Runtime CPU execution route. Exact edge-v3 export, tokenizer/model ABI and runtime compatibility are not verified in this review; a C wrapper is not itself evidence of a speedup.

GLiClass jointly encodes text and labels; do not assume TypeSafe-style independent-question semantics or that scores are invariant when candidate descriptions/order change. No downloaded source/checkpoint hashes or runnable source freeze are claimed here.

### Architectural consequence

Prefer **a bounded deterministic macro skeleton with optional semantic-decision leaves**, rather than an always-on universal local model. Known focus/lease/generation conditions stay deterministic. Do not route those conditions through a classifier. Missing/stale/ambiguous evidence must remain representable without an executable target.

Astra can specialize stable task intent, parameter bindings and branch structure before execution. Current observations, target dependencies and safety events must not be constant-folded into historical truth. Return deterministic matches as typed matches, not fabricated calibrated probabilities. A learned score is a proposal, never current identity proof, task-success evidence, or authority.

Reducing parameter count is one option; reducing repeated semantic work and unnecessary model invocations is another. Neither establishes equal capability to Jev. If the model-free artifact is sufficient for the declared task envelope, there is no requirement to insert or benchmark a neural backend merely to call it System-1.

### Latency/accounting boundary

Record separately: cold authoring/compilation/validation; model/process load if used; current-state acquisition and extraction; serialization/tokenization; decision evaluation; current deterministic admission; input delivery; independently verified effect and release.

The proposed **60 ms target applies initially to warm already-available-state -> typed decision, including local preprocessing/tokenization**, not screenshot-to-effect or whole-task completion. Report the full observation-to-input and effect endpoints separately; do not add component percentiles as though they were the percentile of the total. Never infer fewer frontier calls, money or energy from local throughput alone.

### H / T / D / C / U — prospective discriminator, NOT an activated preregistration

**H:** One bounded repeated typed decision can be handled by the smallest sufficient local artifact, with no task-specific weight updates, warm decision p95 at most 60 ms on a fixed CPU-only batch-1 configuration, and correct UNKNOWN/YIELD behavior. This is a target, not an estimated or measured latency.

**T:** If later activated, first choose one task and one exact candidate, not a model sweep. Compare against a competent model-free macro/rule reference using identical available state, candidates, authority and independent oracle. Change only decision backend. Proposed minimum: 64 unique positive/negative cases plus 1,024 separately identified warm timing calls per backend; timing repetitions are not independent semantic evidence. Pin CPU/threads/clock information, input-length/candidate limits, source/checkpoint/tokenizer/runtime versions, warm-up, schedule and auditor before any formal output. Use a disposable container, fresh allocation IDs, one finite block and first-outcome retention. If the model-free reference is already sufficient, retain that result without requiring the optional model.

**D:** A scoped viability PASS needs correct frozen positive decisions, UNKNOWN/YIELD on the declared negative cases, zero unsafe conversions to executable proposals, unchanged ordinary authority/release requirements, complete accounting, and the latency target. A fast but unsafe result is FAIL. Safe non-completion or missing endpoints is HOLD, not a speed success. Actual frontier-boundary benefit requires separately measured model-in-loop work; it is not established by this local discriminator.

**C:** State extraction may dominate; the smallest encoder may lose necessary semantics; scores may be uncalibrated or candidate-set-sensitive; rule branches may grow too complex; a low-reuse task may not repay compilation/validation; local hardware may be slower than hosted hardware.

**U:** No local measurement, model-size estimate for Jev, CPU latency distribution, calibrated combined uncertainty or coverage factor is available. A bounded text/typed-state result would not establish screenshot perception, real-time guarantees or cross-application transfer. Explicitly record u_c/k as unavailable rather than inventing them.

### Retention and coordination

- #838's first result remains **HOLD_ACTIVE_LABELS_NOT_SUFFICIENT**: only 1/4 replicas passed its full gate. This note neither reruns nor rescues its synthetic sparse-label lane by threshold/tree/query tuning.
- #853 still records **zero formal rows** and an artifact/DNS provisioning failure, not a Mitra runtime/accuracy failure. This text-decision candidate does not reinterpret that evidence or silently replace its frozen question.
- Repository searches for `GLiClass` and open macro PRs found no exact competing result in the returned scope; #857 already owns procedure specialization, so no parallel issue is created. This is not a claim about unindexed/unpublished work.

**Next discriminator:** Does a real repeated decision actually require online semantic inference after Astra has specialized the task, or can a guarded model-free artifact already satisfy its correctness and latency envelope?

**Comment by Unjuno** [comment]
Result update from merged Issue #890 / PR #895 (`e1a94b3b971007405af4e3ca921006f2903a93c7`): **PASS_CHROMIUM_GUARDED_MACRO_COMPILATION_SCOPED**.

Using only the retained #57 persistent Chromium evidence, task1's frontier-authored field-entry procedure was compiled once into a macro parameterized by `{token, field_handle, submit_handle}` with no executable absolute target coordinates and fail-closed current-handle guards. In retained A reuse it matched current targets; at the A→B change it YIELDed before pointer when the old handle was `missing`; after retained repair the same macro admitted current B handles without recompilation. Literal task1 coordinate replay remained apparently applicable because focus/surface/window geometry was unchanged, but mismatched B.

This closes the first mechanism discriminator requested here: procedure semantics can be separated from historical target identity for this scoped field-entry workflow. It does **not** establish general compilation, fresh live task success, or new economics. Per this Issue's sequencing, the next high-information move should be transfer to a second retained procedure/task/domain or a fresh in-envelope/out-of-envelope live check—not adding a learned policy merely because one exists.
